# Data preparation of IMDB movies.
- First, basic overview of the dataset and go through each columns.
- Change the type if necessary.
- Replace or modify particular entries if necessary.
- Make the data makes sense.

In [14]:
import pandas as pd

df = pd.read_csv('imdb_top_1000.csv')

df.head(5)

Poster_Link  \
0  https://m.media-amazon.com/images/M/MV5BMDFkYT...   
1  https://m.media-amazon.com/images/M/MV5BM2MyNj...   
2  https://m.media-amazon.com/images/M/MV5BMTMxNT...   
3  https://m.media-amazon.com/images/M/MV5BMWMwMG...   
4  https://m.media-amazon.com/images/M/MV5BMWU4N2...   

               Series_Title Released_Year Certificate  Runtime  \
0  The Shawshank Redemption          1994           A  142 min   
1             The Godfather          1972           A  175 min   
2           The Dark Knight          2008          UA  152 min   
3    The Godfather: Part II          1974           A  202 min   
4              12 Angry Men          1957           U   96 min   

                  Genre  IMDB_Rating  \
0                 Drama          9.3   
1          Crime, Drama          9.2   
2  Action, Crime, Drama          9.0   
3          Crime, Drama          9.0   
4          Crime, Drama          9.0   

                                            Overview  Meta_score  \
0  Two imprisoned men bond over a number of years...        80.0   
1  An organized crime dynasty's aging patriarch t...       100.0   
2  When the menace known as the Joker wreaks havo...        84.0   
3  The early life and career of Vito Corleone in ...        90.0   
4  A jury holdout attempts to prevent a miscarria...        96.0   

               Director           Star1           Star2          Star3  \
0        Frank Darabont     Tim Robbins  Morgan Freeman     Bob Gunton   
1  Francis Ford Coppola   Marlon Brando       Al Pacino     James Caan   
2     Christopher Nolan  Christian Bale    Heath Ledger  Aaron Eckhart   
3  Francis Ford Coppola       Al Pacino  Robert De Niro  Robert Duvall   
4          Sidney Lumet     Henry Fonda     Lee J. Cobb  Martin Balsam   

            Star4  No_of_Votes        Gross  
0  William Sadler      2343110   28,341,469  
1    Diane Keaton      1620367  134,966,411  
2   Michael Caine      2303232  534,858,444  
3    Diane Keaton      1129952   57,300,000  
4    John Fiedler       689845    4,360,000

In [15]:
# Overview of the data types, I will not necessarily change all the types but it is good to know what we are working with.
df.dtypes

Poster_Link       object
Series_Title      object
Released_Year     object
Certificate       object
Runtime           object
Genre             object
IMDB_Rating      float64
Overview          object
Meta_score       float64
Director          object
Star1             object
Star2             object
Star3             object
Star4             object
No_of_Votes        int64
Gross             object
dtype: object

## Poster_link
- Dropping this column as it does not hold any valuable information.

In [16]:
newdf = df.drop(df.columns[0], axis=1)  # Drop by position
newdf.head(5)

Series_Title Released_Year Certificate  Runtime  \
0  The Shawshank Redemption          1994           A  142 min   
1             The Godfather          1972           A  175 min   
2           The Dark Knight          2008          UA  152 min   
3    The Godfather: Part II          1974           A  202 min   
4              12 Angry Men          1957           U   96 min   

                  Genre  IMDB_Rating  \
0                 Drama          9.3   
1          Crime, Drama          9.2   
2  Action, Crime, Drama          9.0   
3          Crime, Drama          9.0   
4          Crime, Drama          9.0   

                                            Overview  Meta_score  \
0  Two imprisoned men bond over a number of years...        80.0   
1  An organized crime dynasty's aging patriarch t...       100.0   
2  When the menace known as the Joker wreaks havo...        84.0   
3  The early life and career of Vito Corleone in ...        90.0   
4  A jury holdout attempts to prevent a miscarria...        96.0   

               Director           Star1           Star2          Star3  \
0        Frank Darabont     Tim Robbins  Morgan Freeman     Bob Gunton   
1  Francis Ford Coppola   Marlon Brando       Al Pacino     James Caan   
2     Christopher Nolan  Christian Bale    Heath Ledger  Aaron Eckhart   
3  Francis Ford Coppola       Al Pacino  Robert De Niro  Robert Duvall   
4          Sidney Lumet     Henry Fonda     Lee J. Cobb  Martin Balsam   

            Star4  No_of_Votes        Gross  
0  William Sadler      2343110   28,341,469  
1    Diane Keaton      1620367  134,966,411  
2   Michael Caine      2303232  534,858,444  
3    Diane Keaton      1129952   57,300,000  
4    John Fiedler       689845    4,360,000

## View the missing count of the columns

In [17]:
missing_count = newdf.isna().sum()
print(missing_count)

Series_Title       0
Released_Year      0
Certificate      101
Runtime            0
Genre              0
IMDB_Rating        0
Overview           0
Meta_score       157
Director           0
Star1              0
Star2              0
Star3              0
Star4              0
No_of_Votes        0
Gross            169
dtype: int64


## Certificate
- Column that doesn't hold values, I will pass.

In [18]:
temp = newdf.loc[(newdf['Certificate'].isna())] 
# temp

## Released Year
- Type is in object, hence changed it to int
- Checking released year and find the incorrect entries.
- Replaced it.

In [19]:
newdf['Released_Year'].unique()
###
# ['1994' '1972' '2008' '1974' '1957' '2003' '1993' '2010' '1999' '2001'
# '1966' '2002' '1990' '1980' '1975' '2020' '2019' '2014' '1998' '1997'
# '1995' '1991' '1977' '1962' '1954' '1946' '2011' '2006' '2000' '1988'
# '1985' '1968' '1960' '1942' '1936' '1931' '2018' '2017' '2016' '2012'
# '2009' '2007' '1984' '1981' '1979' '1971' '1963' '1964' '1950' '1940'
# '2013' '2005' '2004' '1992' '1987' '1986' '1983' '1976' '1973' '1965'
# '1959' '1958' '1952' '1948' '1944' '1941' '1927' '1921' '2015' '1996'
# '1989' '1978' '1961' '1955' '1953' '1925' '1924' '1982' '1967' '1951'
# '1949' '1939' '1937' '1934' '1928' '1926' '1920' '1970' '1969' '1956'
# '1947' '1945' '1930' '1938' '1935' '1933' '1932' '1922' '1943' 'PG']
### 

# The unique values in the Released_Year column are:
newdf.loc[newdf['Released_Year'] == 'PG']
# The movie Apollo 13 has a Released_Year of PG, which is incorrect. I will replace it with 1970.
newdf.replace('PG', '1970', inplace=True)
newdf.loc[newdf['Series_Title'] == 'Apollo 13']

Series_Title Released_Year Certificate  Runtime  \
966    Apollo 13          1970           U  140 min   

                         Genre  IMDB_Rating  \
966  Adventure, Drama, History          7.6   

                                              Overview  Meta_score  \
966  NASA must devise a strategy to return Apollo 1...        77.0   

       Director      Star1        Star2        Star3        Star4  \
966  Ron Howard  Tom Hanks  Bill Paxton  Kevin Bacon  Gary Sinise   

     No_of_Votes        Gross  
966       269197  173,837,933

In [20]:
newdf['Released_Year'] = newdf['Released_Year'].astype(int)
newdf['Released_Year'].describe()

count    1000.000000
mean     1991.196000
std        23.295026
min      1920.000000
25%      1976.000000
50%      1999.000000
75%      2009.000000
max      2020.000000
Name: Released_Year, dtype: float64

## Series Title
- Manual check requires because anything can be title.
- By looking at each title, there was not an invalid or inappropriate entry.
- While checking, a duplicate was found with different released year.
    - The movie released later years have been renamed from 'Drishyam' to 'Drishyam 2' 

In [21]:
# It's hard to judge and indicate if there are any title is inccorectly entered or not. 
# This requires a manual check.
newdf['Series_Title'].unique() 
newdf['Series_Title'].describe()

# Checking for duplicate entries in the Series_Title column.
duplicates = newdf['Series_Title'][newdf.duplicated('Series_Title', keep=False)]
# print(duplicates)
# Line 87 and 136 have the same Series_Title.

# Checking if they are completely the same entries or not.
newdf.loc[87]
newdf.loc[136]

# They are not the same entries. I will replace the Series_Title of line 136 with the correct title.
newdf.loc[136, 'Series_Title'] = 'Drishyam 2'

newdf['Series_Title'].describe()
# Now there are no duplicate entries in the Series_Title column.

count                         1000
unique                        1000
top       The Shawshank Redemption
freq                             1
Name: Series_Title, dtype: object

## Runtime
- It is an object class with string and int values.
- Convert the type to string and remove string values.
- Change type to integer which is in minutes.
- Lastly, check to see if there is any suspicious entries, such as outliners.

In [22]:
# Runtime is in object type and contains both int and string values.
newdf.sort_values(by='Runtime', ascending=True).head(5)

# - First I will convert the Runtime column to string.
newdf['Runtime'] = newdf['Runtime'].astype(str)

# - Then I will remove the 'min' from the string and convert it to integer.
newdf['Runtime'] = newdf['Runtime'].str.replace(' min', '', regex=True).astype(int)

# newdf.sort_values(by='Runtime', ascending=False).head(5)
newdf['Runtime'].describe() 

# Description of Runtime column looks fine. No outliners.

count    1000.000000
mean      122.891000
std        28.093671
min        45.000000
25%       103.000000
50%       119.000000
75%       137.000000
max       321.000000
Name: Runtime, dtype: float64

## Meta Score
- Missing 157 entries 
- Dropna may affect the entire analysis, hence replace missing values with average Meta Score

In [23]:
# Checking the unique values in the Meta_score column.
newdf['Meta_score'].unique()

# See if incorrect entries are present in the missing Meta_score column.
newdf.loc[newdf['Meta_score'].isna()]

# Replace the missing values with the mean of the Meta_score column.
newdf['Meta_score'].fillna(newdf['Meta_score'].mean(), inplace=True)
newdf['Meta_score'].describe()

/tmp/ipykernel_11444/4232844520.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  newdf['Meta_score'].fillna(newdf['Meta_score'].mean(), inplace=True)


count    1000.00000
mean       77.97153
std        11.36206
min        28.00000
25%        72.00000
50%        77.97153
75%        85.25000
max       100.00000
Name: Meta_score, dtype: float64

## Gross
- Based on the given set, half of the missing Gross values released are before 1974.
- There are so many factors to be considered if wanted to find an individual gross value for each movie.
    - Such as: Movies played in which countries, currency, inflation, and how many people watched the movie.
- This case requires more time and studies.
- Precision is key and important when doing data analysis.

### This maybe a pontential fix, but does not gaurantee to be correct at all.
- I would personally not use this method: 

```py
# Group by dacades and replace with median values.
newdf['Decade'] = (newdf['Released_Year'] // 10) * 10  # Create decade column  
median_gross_by_decade = newdf.groupby('Decade')['Gross'].median()  
newdf.loc[newdf['Gross'].isna(), 'Gross'] = newdf['Decade'].map(median_gross_by_decade)
```

In [24]:
# Checking the unique values in the Gross column.
newdf['Gross'].unique()

# See if incorrect entries are present in the missing Gross column.
newdf['Gross'].describe()

# Convert the Gross column to string and replace the missing values with 0.
newdf['Gross'] = newdf['Gross'].astype(str)

# Replace the missing values with 0.
newdf['Gross'].replace('nan', '0',inplace=True)

# Double checking if the missing values are replaced with 0.
newdf['Gross'].describe()
newdf.loc[newdf['Gross'] == '0'].describe()

/tmp/ipykernel_11444/3501585163.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  newdf['Gross'].replace('nan', '0',inplace=True)


Released_Year     Runtime  IMDB_Rating  Meta_score    No_of_Votes
count     169.000000  169.000000   169.000000  169.000000     169.000000
mean     1977.952663  117.053254     7.960947   80.241635   69350.745562
std        28.360312   30.254921     0.234295    8.085057   57631.153579
min      1920.000000   64.000000     7.600000   51.000000   25312.000000
25%      1955.000000   96.000000     7.800000   77.971530   33656.000000
50%      1974.000000  111.000000     8.000000   77.971530   51913.000000
75%      2006.000000  129.000000     8.100000   85.000000   73992.000000
max      2020.000000  321.000000     8.600000  100.000000  405801.000000

## End of Data Cleaning
- For a personal note, it seems that the dataset has been cleaned and ready to be analyzed.

In [25]:
# Lastly, check the missing values in the dataset.
missing_count = newdf.isna().sum()
print(missing_count)
# Other than certificate, it looks like there are no missing values in the dataset.

Series_Title       0
Released_Year      0
Certificate      101
Runtime            0
Genre              0
IMDB_Rating        0
Overview           0
Meta_score         0
Director           0
Star1              0
Star2              0
Star3              0
Star4              0
No_of_Votes        0
Gross              0
dtype: int64


In [26]:
# Export the cleaned data to a new csv file.
newdf.to_csv('cleaned_imdb_top_1000.csv', index=False)